File: heavy-metal-bake-sale/csharp/src/HeavyMetalBakeSale/Money.cs
namespace HeavyMetalBakeSale;

public readonly record struct Money(decimal Amount)
{
    public static Money Zero => new(0m);

    public bool IsNonNegative => Amount >= 0m;

    public static Money operator +(Money a, Money b) => new(a.Amount + b.Amount);
    public static Money operator -(Money a, Money b) => new(a.Amount - b.Amount);
    public static bool operator >(Money a, Money b) => a.Amount > b.Amount;
    public static bool operator <(Money a, Money b) => a.Amount < b.Amount;
    public static bool operator >=(Money a, Money b) => a.Amount >= b.Amount;
    public static bool operator <=(Money a, Money b) => a.Amount <= b.Amount;

    public string ToDisplay() => $"${Amount:F2}";
}

File: heavy-metal-bake-sale/typescript/src/Money.ts
export class Money {
  static readonly zero = new Money(0);

  constructor(public readonly amount: number) {}

  plus(other: Money): Money {
    return new Money(Math.round((this.amount + other.amount) * 100) / 100);
  }

  minus(other: Money): Money {
    return new Money(Math.round((this.amount - other.amount) * 100) / 100);
  }

  lessThan(other: Money): boolean {
    return this.amount < other.amount;
  }

  equals(other: Money): boolean {
    return this.amount === other.amount;
  }

  toDisplay(): string {
    return `$${this.amount.toFixed(2)}`;
  }
}
